crates/wasm-mutate/src/mutators/peephole/mod.rs

```diff
@@ -119,6 +173,57 @@ impl Mutator for PeepholeMutator {
         Ok(module)
     }
 
+    /// Condition to apply the unfold operator
+    /// check that the var is a constant
+    /// Condition to apply the unfold operator
+    /// check that the var is a constant
+    fn is_const(&self, vari: &'static str) -> impl Fn(&mut EG, Id, &Subst) -> bool {
+        move |egraph: &mut EG, _, subst| {
+            let var = vari.parse();
+            match var {
+                Ok(var) => {
+                    let eclass = &egraph[subst[var]];
+                    if eclass.nodes.len() == 1 {
+                        println!("{:?}", eclass);
+                        let node = &eclass.nodes[0];
+                        match node {
+                            Lang::I32Const(_) => true,
+                            _ => false,
+                        }
+                    } else {
+                        false
+                    }
+                }
+                Err(_) => false,
+            }
+        }
+    }
+}
+
+/// Meta mutator for peephole
+impl Mutator for PeepholeMutator {
+    fn mutate(
+        &self,
+        config: &crate::WasmMutate,
+        rnd: &mut rand::prelude::SmallRng,
+        info: &mut crate::ModuleInfo,
+    ) -> Result<Module> {
+        let rules: &[Rewrite<Lang, PeepholeMutationAnalysis>] = &[
+            rewrite!("unfold-2";  "?x" => "(unfold ?x)" if self.is_const("?x") ), // Use a custom instruction-mutator for this
+            // This specific rewriting rule has a condition, it should be appplied if the operand is a constant
+            // To do so we can write all symbols representing constants as ?c when we translate wasm to eterm
+            rewrite!("strength-undo";  "(i32.shl ?x 1)" => "(i32.mul ?x ?x)"),
+            rewrite!("strength-undo1";  "(i32.shl ?x 2)" => "(i32.mul ?x 2)"),
+            rewrite!("strength-undo2";  "(i32.shl ?x 3)" => "(i32.mul ?x 8)"),
+            rewrite!("add-1";  "(i32.add ?x ?x)" => "(i32.mul ?x 2)"),
+            rewrite!("idempotent-1";  "?x" => "(i32.or ?x ?x)"),
+            rewrite!("idempotent-2";  "?x" => "(i32.and ?x ?x)"),
+            rewrite!("mem-load-shift";  "(i.load ?x)" => "(i.load (i32.add ?x rand))"), // Check why this is generating a lot of the same replacements
```

Comment:
This doesn't seem like a valid rewrite to me

Reply:
Yeah...the mem-load-shift is not semantically correct, I will add the corresponding filter coming from the configuration